r/tests/testthat/helper-parquet.R

```diff
@@ -0,0 +1,24 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+expect_parquet_roundtrip <- function(tab, ...) {
+  tf <- tempfile()
+  on.exit(unlink(tf))
+
+  write_parquet(tab, tf, ...)
+  expect_equal(read_parquet(tf, as_data_frame = FALSE), tab)
```

Comment:
I like the idea here but there are a couple of problems: 

1. `Table__Equals` is not implemented:
```r
> tab <- Table$create(a=1, b=4)
> tab2 <- Table$create(c=3:4, d=5:6, h=8:9)
> tab == tab2
Error in Table__Equals(self, other) : 
  could not find function "Table__Equals"
```

2. `testthat::expect_equal` doesn't do `x == y`, it does `all.equal(x, y)`: https://github.com/r-lib/testthat/blob/master/R/compare.R#L48 

And because of how `all.equal.environment` works, any of our R6 objects of the same class will compare as equal:

```
> all.equal(tab, tab2)
[1] TRUE
```

So a better test would be `expect_true(read_parquet(tf, as_data_frame = FALSE) == tab)`, provided that `Table__Equals` is implemented.

(Aside/instead: we should probably add `all.equal.Object <- function(x, y, ...) x == y` as a fallback so that `expect_equal` does what we expect it to in tests. I observed this issue on a branch I started involving schema because I found that tests I expected to fail because I hadn't implemented the fix yet were not failing.)

Reply:
Thanks. I like the aside too, I'll put that in. It's odd I was certain I added `__Equals` and it's trivial anyway. ☕️